### PR TITLE
Editable: implement TinyMCE on the core Editable component

### PR DESCRIPTION
**Pull request by youknowriad**
In this PR, I'm setting up TinyMCE in the `Editable` component to ease building blocks using this generic component. Details on this component on the Doc PR https://github.com/WordPress/gutenberg/pull/302/files#diff-8973b2df4ecb1bd89ade51059ee90658R157

**Notes**
- I'm reusing the WordPress TinyMCE scripts (v3)
- I had to expose the React `Component` class, I wonder if this should be available for block creators or not.
- For now, the editable component is very basic. It takes a `value` prop and triggers `onChange` callback on Focus out (to avoid cursor jumping).
 - With all this `.bind` calls, I felt the need for the class properties babel plugin.

### CODE REVIEW

**Comment by youknowriad** [comment]
@aduth The tests are failing because The `editable` component is being loaded and it requires the `wp` globals.

 - I personally think we should recreate the `registration.js` file to isolate the block registration for the other unrelated stuff and this should resolve this issue by not loading the `editbale` component while testing. 

 - We can also mock the globals in `bootstrap-test.js`

Thoughts? @nylen

File: index.php [hunk 1]
```diff
@@ -40,8 +40,9 @@ function gutenberg_register_scripts() {
 	wp_register_script( 'react-dom', 'https://unpkg.com/react-dom@15/dist/react-dom' . $suffix . '.js', array( 'react' ) );
 
 	// Editor
+	wp_register_script( 'tinymce_js', includes_url( 'js/tinymce/' ) . 'wp-tinymce.php', array( 'jquery' ) );
```
Comment (mtias) [inline index.php]: I think we should get the trunk version, which has the link boundaries.

**Comment by aduth** [comment]
>this should resolve this issue by not loading the `editable` component while testing.

Sounds more like a workaround than a proper fix.

File: blocks/components/editable/index.js [hunk 2]
```diff
@@ -1,3 +1,98 @@
-export default function Editable( { value } ) {
-	return wp.element.createElement( 'p', null, value );
+const el = wp.element.createElement;
+let editorInstance = 1;
+
+export default class Editable extends wp.element.Component {
+	constructor( ...args ) {
+		super( ...args );
```
Comment (aduth) [inline blocks/components/editable/index.js]: Minor: `...arguments` works just as well and might (a) more clearly indicate you're not really using the arguments and (b) avoid needing to update the call if ever you decide to use one of the arguments.

File: blocks/components/editable/index.js [hunk 3]
```diff
@@ -1,3 +1,98 @@
...
+	}
+
+	updateContent() {
+		let bookmark;
+		if ( this.isFocused ) {
+			bookmark = this.editor.selection.getBookmark( 2, true );
+		}
+		this.editor.setContent( this.props.value );
+		if ( this.isFocused ) {
+			this.editor.selection.moveToBookmark( bookmark );
```
Comment (aduth) [inline blocks/components/editable/index.js]: Out of curiosity, does it make any difference if we were to call this while the editor does not have focus?

File: blocks/components/editable/index.js [hunk 4]
```diff
@@ -1,3 +1,98 @@
...
+
+	componentWillUnmount() {
+		if ( this.editor ) {
+			this.editor.destroy();
+		}
+	}
+
+	componentDidUpdate( prevProps ) {
+		if ( this.props.value !== prevProps.value ) {
+			this.updateContent( !! prevProps.focusConfig );
```
Comment (aduth) [inline blocks/components/editable/index.js]: We're not using the argument passed?

File: blocks/components/editable/index.js [hunk 5]
```diff
@@ -1,3 +1,98 @@
...
+		}
+	}
+
+	render() {
+		return el( 'div', {
+			id: this.id,
+			contentEditable: true
```
Comment (aduth) [inline blocks/components/editable/index.js]: Shouldn't need to assign this, as TinyMCE will do so on our behalf.

File: blocks/components/editable/index.js [hunk 6]
```diff
@@ -1,3 +1,98 @@
...
+	}
+
+	componentDidUpdate( prevProps ) {
+		if ( this.props.value !== prevProps.value ) {
+			this.updateContent( !! prevProps.focusConfig );
+		}
+	}
+
+	render() {
+		return el( 'div', {
```
Comment (aduth) [inline blocks/components/editable/index.js]: JSX?

File: blocks/components/editable/index.js [hunk 7]
```diff
@@ -1,3 +1,98 @@
...
+		editor.on( 'focusout', this.onFocusOut );
+	}
+
+	onInit() {
+		this.editor.setContent( this.props.value );
+	}
+
+	onChange() {
+		const value = this.editor.getContent();
+		if ( value === this.props.value ) {
```
Comment (aduth) [inline blocks/components/editable/index.js]: - Would [`isDirty`](https://www.tinymce.com/docs/api/tinymce/tinymce.editor/#isdirty) work here?
- Alternatively, would specifying `format: raw` to `getContent` (much more performant) save us at all, or would we expect most calls to this function to pass the condition anyways?

File: editor/blocks/text-block/index.js [hunk 8]
```diff
@@ -1,11 +1,11 @@
-const { query, html } = wp.blocks.query;
+const { html } = wp.blocks.query;
 
 wp.blocks.registerBlock( 'wp/text', {
 	title: 'Text',
 	icon: 'text',
 
 	attributes: {
-		value: query( 'p', html() )
+		value: html()
```
Comment (aduth) [inline editor/blocks/text-block/index.js]: Might still be value in specifying `'p'` argument to at least limit value to text within the `<p>` element of the block?

File: element/index.js [hunk 9]
```diff
@@ -28,3 +28,8 @@ export function createElement( type, props, ...children ) {
 export function render( element, target ) {
 	reactRender( element, target );
 }
+
+/**
+ * A base class to create WordPress Components (Refs, state and lifecycle hooks)
+ */
+export const Component = ReactComponent;
```
Comment (aduth) [inline element/index.js]: Prefixed imports bothering me a bit. We could instead do...

```js
import { Component } from 'react';

// ...
export { Component };
```

What do you think?

File: blocks/components/editable/index.js [hunk 10]
```diff
@@ -1,3 +1,98 @@
...
+export default class Editable extends wp.element.Component {
+	constructor( ...args ) {
+		super( ...args );
+		this.onInit = this.onInit.bind( this );
+		this.onSetup = this.onSetup.bind( this );
+		this.onChange = this.onChange.bind( this );
+		this.onFocusIn = this.onFocusIn.bind( this );
+		this.onFocusOut = this.onFocusOut.bind( this );
+		this.id = `tinymce-instance-${ editorInstance }`;
+		editorInstance++;
```
Comment (aduth) [inline blocks/components/editable/index.js]: Doesn't make a whole lot of difference, but we could assign the incremented value as a static property of the class:

```js
export default class Editable extends wp.element.Component {
	static instance = 1;

	constructor() {
		// ...
		this.id = `tinymce-instance-${ this.constructor.instance }`;
		this.constructor.instance++;

		// or: (increment slightly less obvious)
		this.id = `tinymce-instance-${ this.constructor.instance++ }`;
	}
```

**Comment by nylen** [comment]
> We can also mock the globals in `bootstrap-test.js`

This, or mocking the `editable` component itself, would be my preference.  I'd rather avoid structuring the code with a fairly arbitrary division imposed by the test setup.

Also worth thinking about how we will write full-browser tests including the `editable` component.

File: blocks/components/editable/index.js (hunk 10 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: This will be removed since I'm updating to Tiny4 which allows us to use the dom node instead of generating a random id  :)

File: blocks/components/editable/index.js (hunk 7 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: Using `isDirty` coupled with `save` is possible I think 👍
And yeah most calls will pass the condition here, I think

File: blocks/components/editable/index.js (hunk 3 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: It can cause the cursor to jump to the beginning of the content editable. I think right now, it does not happen because we're not updating the content from outside the component yet.
I've had some hard time fixing these kind of bugs on the `per-block` prototype.

File: editor/blocks/text-block/index.js (hunk 8 above)
Comment (youknowriad) [inline editor/blocks/text-block/index.js]: This is a good question. If we limit the value to the text within the `p`, we should limit the `Editable` to create a unique `p` and the `onChange` callback of the `Editable` component should extract the content of the `p` before calling the `onChange` prop.

This is probably what we want, I was thinking we could leave this to a separate PR.

File: element/index.js (hunk 9 above)
Comment (youknowriad) [inline element/index.js]: I'm ok either way :)

**Comment by youknowriad** [comment]
> This, or mocking the editable component itself, would be my preference.

Not sure how to do that.

>  I'd rather avoid structuring the code with a fairly arbitrary division imposed by the test setup.

This is true, but I was thinking we should move the registration API to its own file regardless of the way we fix the test setup

File: blocks/components/editable/index.js (hunk 3 above)
Comment (aduth) [inline blocks/components/editable/index.js]: Yeah my thinking was if the editor wasn't focused anyways, setting content and moving bookmark wouldn't have any impact on the current active focus. I did some basic testing and this appears to be the case? I was entering this in my console at page load:

```js
var bookmark = tinymce.editors[ 1 ].selection.getBookmark( 2, true );
tinymce.editors[ 1 ].setContent( tinymce.editors[ 0 ].getContent() );
tinymce.editors[ 1 ].selection.moveToBookmark( bookmark );
```

File: bootstrap-test.js [hunk 11]
```diff
@@ -13,3 +13,8 @@ global.document = require( 'jsdom' ).jsdom( '', {
 } );
 global.window = document.defaultView;
 global.navigator = window.navigator;
+global.wp = {
```
Comment (aduth) [inline bootstrap-test.js]: The issue seems to be that the globals are only defined in the main bundle configuration, whereas the entry point of the tests bundle includes only test files. I might toy with the configuration a bit to see if I can fix it so we don't need to stub these out.

**Comment by aduth** [comment]
Should we remove `contenteditable` from main wrapper div at this point?

File: blocks/components/editable/index.js (hunk 3 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: You're right! Can't remember exactly why I had those on the `per-block` editor. Removing these checks for now

**Comment by aduth** [comment]
In 97ba143 I reconfigured Webpack to ensure that globals from each submodule are exposed for the tests. Unfortunately this means pulling in React again separate from the one we define in the plugin, since in the context of the Node tests it can't pull from the CDN like it does in the browser. We might want to think of a better way to avoid managing two separate copies of these vendor dependencies.

**Review by aduth — APPROVED** [review]
Looks good 👍

File: blocks/components/editable/index.js [hunk 12]
```diff
@@ -1,3 +1,74 @@
...
+		this.onChange = this.onChange.bind( this );
+		this.bindNode = this.bindNode.bind( this );
+	}
+
+	componentDidMount() {
+		this.initialize();
+	}
+
+	initialize() {
+		const config = {
```
Comment (iseulde) [inline blocks/components/editable/index.js]: To recreate default `contenteditable` behaviour, we should probably add `browser_spellcheck: true`.

File: editor/blocks/text-block/index.js (hunk 8 above)
Comment (joyously) [inline editor/blocks/text-block/index.js]: How does specifying `p` here affect `wpautop()` type of stuff?

Comment (youknowriad) [inline editor/blocks/text-block/index.js]: I think this is unrelated to `wpautop`.

The `wpautop` needs to be implemented by catching the `Enter` key press and checking whether we should create a new block after the current one or not. (Similar behaviour can be seen on the `per-block`  prototype)

Comment (joyously) [inline editor/blocks/text-block/index.js]: Well, sort of. What I was interested in was what is actually saved. If I look at the post content of my current blog, there are a lot of text paragraphs separated by one blank line. There are not actual `p` tags. Somewhere along the way to display `wpautop` puts `p` tags? Will the new editor save the `p` tag that the old editor did not?

File: blocks/components/editable/index.js [hunk 13]
```diff
@@ -1,3 +1,75 @@
...
+
+	updateContent() {
+		const bookmark = this.editor.selection.getBookmark( 2, true );
+		this.editor.setContent( this.props.value );
+		this.editor.selection.moveToBookmark( bookmark );
+	}
+
+	componentWillUnmount() {
+		if ( this.editor ) {
+			this.editor.destroy();
```
Comment (spocke) [inline blocks/components/editable/index.js]: Use this.editor.remove(); destroy is more for memory cleanups on unload of the page not really a problem now days on modern browsers. Destroy is called by remove.

File: blocks/components/editable/index.js [hunk 14]
```diff
@@ -1,3 +1,75 @@
...
+			}
+		};
+
+		tinymce.init( config );
+	}
+
+	onSetup( editor ) {
+		this.editor = editor;
+		editor.on( 'init', this.onInit );
+		editor.on( 'focusout', this.onChange );
```
Comment (spocke) [inline blocks/components/editable/index.js]: focusout will when focus is shifted internally within nested contenteditables. Use "blur" instead or possible 'change' since that is called on every undo level add including 'blur' since that adds an undo level. Might be to frequent though.

File: editor/blocks/text-block/index.js (hunk 8 above)
Comment (youknowriad) [inline editor/blocks/text-block/index.js]: I'm assuming we're talking about the visual editor. Then yes, in its current form this block saves the content with the `p` tag. A block declares its saved markup in the `save` function. You can see that the current `wp/text` block wraps the content with a `p` tag.

But this is far from being the final version of the text block, it's just here to serve as an example of block while we're building the "surroundings", how we parse the content, how we save it back, how we render the editables etc...

Comment (joyously) [inline editor/blocks/text-block/index.js]: Yes, it's an example, so I would hope you "do it right". If it saves the `p` to the content (where the old editor did not), does that affect `wpautop`? 
I get that "it's not the final version" answer a lot, but my experience is that those things are then forgotten and harder to change later. I'm just trying to think of consequences down the line, instead of "building this block now".

Comment (youknowriad) [inline editor/blocks/text-block/index.js]: This is the exact reason why we're trying to build this as a plugin and to tie it to real data (real posts), for now, my answer to this is "I don't know". We'll have to try the editor in the context of real data to figure out these consequences (and fix all the unwanted behaviour).

File: blocks/components/editable/index.js (hunk 14 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: Since this is a "controlled" TinyMCE (each time the content prop changes, we update the content), I had issues while using `change` on the `per-block` prototype. I haven't tried `blur` though
